design/pipeline-caching.md

```diff
@@ -0,0 +1,366 @@
+
+Caching can introduce many complexities into build pipelines and our goal is to strike the balance between "correct" and "good enough" where "good enough" allows us to provide performance benefits without a disproporionate amount of effort on behalf of the user.
+
+## Proposal
+
+This proposal contains a number of interelated elements. I've tried to build it up sequentially but sometimes you'll need to read further along to understand why something is the way it is.
+
+### The basics & task integration
+
+When work is done on an agent as part of a pipeline, it is represented as a _job_ when a set of _steps_ which are implemented by tasks. For many pipelines the YAML syntax obscures the underlying "tasky" nature of Azure Pipelines, but they are there nonetheless. As a result I'm starting with how pipeline caching will integrate with the underlying task infrastructure.
+
+Detecting a cache hit or miss will be performend by the agent worker. As the worker prepares to execute each task it will check to see whether any given task has caching enabled and if so it will determine the key and check for a cache hit. If a hit is detected it will set about restoring the cached content into the pipeline execution environment. Exactly how that content is restored will be discussed a bit later.
+
+The benefit of this approach is that caching can be applied to all existing tasks without any additional work by the task author. Support for caching can be enhanced to provide sensible defaults (e.g. what files to consider for cache key input).
+
+Customers that are using the task-designer experience will be able to take advantage of pipeline caching. Cache control properties will be available in a _Caching_ group under each task in the designer.
+
+![Picture of Azure Pipelines task designer with Caching group](images/pipeline-caching-task-editor.jpg)
+
+From this screenshot you can see five inputs to the caching infrastructure:
+
+1. **Use cache flag**; controls whether the cache is used for this task in the case of a cache hit.
+2. **Populate cache flag**; controls whether this task is capable of populating the cache.
+3. **Key expression**; an expression that ultimately returns a string-based key which forms the key to identify the cache contents. May be combined with other scoping information before use.
+4. **Capture expression**; an expression that returns a set of things to capture. The most common case would be files, but it may also be necessary to capture things like environment variables etc which might be the output of a task.
+5. **Strategy**; an option that controls how cached content will be stored. For large artifacts our dedup protocol might be most appropriate, in other cases using a zip file might be more efficient.
+
+#### Cache use vs. population
```

Comment:
This makes sense though it should be controlled by simply using conditionals to not run the cache restore step in whatever cases you require.